Improve editor state persistence

Geometror opened the pull request:
This PR is an attempt to improve the editor state preservation when restarting the editor. Strangely I couldn't find any GH issues on this topic, but I know this has been a inconvenience for quite a few users, myself included :)

The following settings are now correctly saved and loaded:
- Active open scene
- Bottom panel:
	- Height (split offset) offset
	- Active tab (e.g. debugger, shader editor)
	- Shader editor:
		- Open shaders (incl. VisualShader)
		- Selected(currently editing) shader
	- Active debugger tab
- Active tab per dock
- Active main editor plugin (2D/3D/Script/Asset/...)
- Script editor: 
	- Selected(currently editing) script
- Selected nodes per scene
- Some viewport specific settings which were not correctly preserved (mostly regressions)
	- Orthogonal/perspective view 
	- Advanced display (debug) modes (e.g. normal buffer, GI buffer, shadow map)
	- Half resolution
- FileSystem dock
   - selected paths
   - expanded folders

Additional misc changes (internal, not compat-breaking):
- Rename key `"use_orthogonal"` to `"orthogonal"`
- Rename `VIEW_DISPLAY_SHADELESS` to `VIEW_DISPLAY_UNSHADED`
- Some renames of internal methods and variables to better describe what they do

While this PR should fix the most annoying cases where the layout/settings would be lost, there are almost certainly still some left. Please let me know if you find one that should be included in this PR :)

Somewhat related side note: I realized that after 4.0 is out we need to do some significant internal refactoring of some core editor classes/files (e.g. editor_node/node_3d_editor_plugin). There's lots of duplicated/legacy code, confusing/short variable names, etc. which make it very hard to maintain them.

And lastly, since some visual comparisons won't do any harm:

---

**Editor before and after the restart with this PR**
![image](https://user-images.githubusercontent.com/50084500/215292294-18656e8e-9aeb-492b-8f68-19da5c43b1ee.png)

**Editor after restart with beta 16 (manually switched from scene "2d_test" to "main_test")**
![image](https://user-images.githubusercontent.com/50084500/215291491-4c8acbcd-4c93-4dab-923c-77e2f2242c11.png)

*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/4789

Paulb23 commented on editor/plugins/script_editor_plugin.cpp:
```diff
@@ -3222,6 +3230,10 @@ void ScriptEditor::get_window_layout(Ref<ConfigFile> p_layout) {
 	}
 
 	p_layout->set_value("ScriptEditor", "open_scripts", scripts);
+	PackedInt32Array selected_scripts = script_list->get_selected_items();
```
Would it not be better to store the resource path, not sure it's guaranteed to load in the same order / external modification? If the file is moved / deleted we wouldn't load it here anyway.

Geometror replied:
Do you know what's the sense behind of ScriptEditor::set_scene_root_script ? I noticed that it always resets the script to the first in the list (while the scene root hasn't even a script attached to it). Besides, in my opinion opening the scene root script is not very user-friendly.

Paulb23 replied:
It's an very old feature. If I can remember correctly, it used to be quite useful as originally script and scene editing were tied together, so it would automatic open the roots node's script. Can't say how many use it now.

Paulb23 commented on editor/plugins/shader_editor_plugin.cpp:
```diff
@@ -196,6 +196,57 @@ VisualShaderEditor *ShaderEditorPlugin::get_visual_shader_editor(const Ref<Shade
 }
 
+void ShaderEditorPlugin::set_window_layout(Ref<ConfigFile> p_layout) {
+	if (!p_layout->has_section("ShaderEditor")) {
```
This should align with `script_editor` and check `EDITOR_GET("text_editor/behavior/files/restore_scripts_on_load")`

Geometror replied:
I'm not sure about this. It might be misleading for users that this setting also affects the shader editor, in which you can edit both textual and visual shaders. Maybe we should provide a separate editor flag?

Paulb23 replied:
I guess it depends, I would class shaders as scripts. Either a new or sharing that setting is fine with me.

Paulb23 reviewed (commented):
> Strangely I couldn't find any GH issues on this topic

The `ShaderEditor` is tracked by #29094. Though, that would include caching the full sate i.e bookmarks, carets etc that the `ScriptEditor` currently does.

Iniquitatis commented on editor/plugins/node_3d_editor_plugin.h:
```diff
@@ -124,25 +124,28 @@ class Node3DEditorViewport : public Control {
 		VIEW_DISPLAY_LIGHTING,
+		VIEW_DISPLAY_UNSHADED,
 		VIEW_DISPLAY_ADVANCED,
+		// Advanved menu:
```
```suggestion
		// Advanced menu:
```

Geometror commented:
Updated the PR with the following changes.

- For the current scene, selected script and selected shader the resource path is now stored instead of the tab index. (as suggested by @Paulb23)
- Add editor setting `"editors/shader_editor/behavior/files/restore_shaders_on_load"` (default: `true`)
- Add editor setting `"text_editor/behavior/files/select_scene_root_script_on_load"` (default: `false`) [This might require some discussion. Do you find this behavior useful?]

michaelhue commented:
It would be great if the FileSystem panel would preserve expanded folders.

Geometror commented:
Rebased and updated:
- Fixed a crash
- Improved state preservation for the FileSystemDock
	- selected paths
	- expanded folders (as suggested by @michaelhue)

Geometror commented:
- Fixed a bug that caused the selected nodes not to be restored when the scene's root node was a canvas item.
- Some internal renames of unclearly named methods

KoBeWi commented on doc/classes/EditorSettings.xml:
```diff
@@ -763,6 +766,9 @@
 			If [code]true[/code], reopens scripts that were opened in the last session when the editor is reopened on a given project.
 		</member>
+		<member name="text_editor/behavior/files/select_scene_root_script_on_load" type="bool" setter="" getter="">
+			If [code]true[/code], selects the root node's script in the script editor when opening a scene.
```
Isn't this the same as `open_dominant_script_on_scene_change`?

Geometror replied:
Actually yes, somehow I managed to overlook `open_dominant_script_on_scene_change`. Removed the new setting. 
Regarding the setting itself: Personally I think this should be turned off by default. I find it more useful when the last script I worked on is restored. What do you think? Maybe we should make a poll?

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -3747,8 +3746,16 @@ Error EditorNode::load_scene(const String &p_scene, bool p_ignore_broken_deps, b
+	editor_state_cf.instantiate();
+	Error editor_state_cf_err = editor_state_cf->load(config_file_path);
+	if (editor_state_cf_err == OK || editor_state_cf->has_section("editor_states")) {
+		// Must not exist.
```
I know this comment was here, but it's rather odd. Maybe it could be changed to something clearer like `Load states if found.` or removed, because this code is self-explanatory.

Geometror replied:
Removed it

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -3769,8 +3776,20 @@ Error EditorNode::load_scene(const String &p_scene, bool p_ignore_broken_deps, b
+		TypedArray<NodePath> selected_node_list = editor_state_cf->get_value("editor_states", "selected_nodes", TypedArray<String>());
+
+		for (int i = 0; i < selected_node_list.size(); i++) {
+			Node *selected_node = new_scene->get_node(selected_node_list[i]);
```
```suggestion
			Node *selected_node = new_scene->get_node_or_null(selected_node_list[i]);
```
This avoids potential errors.

KoBeWi commented on editor/editor_node.h:
```diff
@@ -646,13 +646,16 @@ class EditorNode : public Node {
 	void _update_dock_slots_visibility(bool p_keep_selected_tabs = false);
 	void _dock_tab_changed(int p_tab);
 
+	void _save_central_editor_layout_to_config(Ref<ConfigFile> p_layout, const String &p_section);
+	void _load_central_editor_layout_from_config(Ref<ConfigFile> p_layout, const String &p_section);
```
```suggestion
	void _save_central_editor_layout_to_config(Ref<ConfigFile> p_layout);
	void _load_central_editor_layout_from_config(Ref<ConfigFile> p_layout);
```
I know it's for consistency, but it's pointless if it ever uses only one section.

Geometror replied:
Cleaned these methods up, added a new constant in editor_node.cpp `EDITOR_NODE_CONFIG_SECTION` and replaced all occurences of the string literal "EditorNode" with it.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4784,6 +4804,31 @@ void EditorNode::_save_docks_to_config(Ref<ConfigFile> p_layout, const String &p
+	// BFS to save the collapsed state of the tree.
+	TreeItem *root_ti = FileSystemDock::get_singleton()->get_tree_control()->get_root();
+	List<TreeItem *> queue;
+	Vector<String> collapsed_paths;
```
```suggestion
	Vector<String> uncollapsed_paths;
```
Typo?

Geometror replied:
Yes :)

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -4784,6 +4804,31 @@ void EditorNode::_save_docks_to_config(Ref<ConfigFile> p_layout, const String &p
+	while (queue.size()) {
+		TreeItem *ti = queue.back()->get();
+		queue.pop_back();
+		if (!ti->is_collapsed()) {
```
This condition is not enough, because it will store also files. You need to check if it's a directory first.

Geometror replied:
You're right. While it basically works since the collapsed property is ignored for leaves, just saving this information for inner nodes of the tree is more efficient (especially when several hundred files are involved). Instead of checking directlity whether it's a directory, I went with this general approach which might be a bit more future-proof and cleaner:
```suggestion
		if (!ti->is_collapsed() && ti->get_child_count() > 0) {
```

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -5035,7 +5182,15 @@ void EditorNode::_load_open_scenes_from_config(Ref<ConfigFile> p_layout, const S
 	}
-	save_layout();
+
+	// The canvas item editors clears the selection in set current scene for some reason!!!!
```
```suggestion
	// The canvas item editor clears the selection in set current scene for some reason.
```
Although I'd check if this is still true, because there were some editor changes that could fix this.

Geometror replied:
Ooops, this was actually a note to myself that should be removed

KoBeWi commented on scene/gui/tree.cpp:
```diff
@@ -4990,6 +4990,26 @@ TreeItem *Tree::get_item_with_text(const String &p_find) const {
 }
 
+TreeItem *Tree::get_item_with_metadata(const Variant &p_find, int p_column) const {
+	if (p_column < 0) {
```
I think this is reversed? Column greater or equal 0 is a specific one.

Geometror replied:
🤦‍♂️ Fixed.

KoBeWi commented on editor/plugins/node_3d_editor_plugin.cpp:
```diff
@@ -3896,19 +3897,27 @@ Dictionary Node3DEditorViewport::get_state() const {
+
+	// Find selected display mode.
+	int display_mode = VIEW_DISPLAY_NORMAL;
+	for (int i = VIEW_DISPLAY_NORMAL; i <= VIEW_DISPLAY_UNSHADED; i++) {
```
If you are iterating, you should add `VIEW_DISPLAY_MAX` constant.

Geometror replied:
It's a bit of a mess here because I need to split the iteration due to the advanced debug view menu. The enum already contains a VIEW_MAX which marks the end of the whole enum. However, there is also already VIEW_DISPLAY_ADVANCED which separates both menus (and has the same function as the VIEW_DISPLAY_MAX constant you suggested), so I changed it to:
```suggestion
	for (int i = VIEW_DISPLAY_NORMAL; i < VIEW_DISPLAY_ADVANCED; i++) {
```

KoBeWi commented:
Seems like 3D editor state is not restored correctly. No matter what I set, I start with single viewport with default options.

Geometror commented:
@KoBeWi Adressed your review comments.
The regression regarding the 3D editor has been fixed, it was introduced by accidently moving the call of `set_edited_scene(new_scene);` after `_load_editor_plugin_states_from_config(editor_state_cf);` during restructuring.

KoBeWi commented:
There's one more bug with filesystem dock. It stores `<null>` in uncollapsed paths (probably from `res://`).

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -156,6 +157,8 @@ EditorNode *EditorNode::singleton = nullptr;
 // The metadata key used to store and retrieve the version text to copy to the clipboard.
 static const String META_TEXT_TO_COPY = "text_to_copy";
 
```
```suggestion
```

KoBeWi commented on editor/plugins/node_3d_editor_plugin.cpp:
```diff
@@ -3896,19 +3897,27 @@ Dictionary Node3DEditorViewport::get_state() const {
+			break;
+		}
 	}
+	for (int i = VIEW_DISPLAY_DEBUG_PSSM_SPLITS; i <= VIEW_DISPLAY_MOTION_VECTORS; i++) {
```
```suggestion
	for (int i = VIEW_DISPLAY_ADVANCED + 1; i < VIEW_DISPLAY_MAX; i++) {
```
This would make it less hard-coded, but I'm not sure if it's worth it 🤔
(`VIEW_DISPLAY_MAX` is a new constant)

Geometror replied:
I like this idea, changed it.

Geometror commented:
Okay, rewrote a few things.
All problems you mentioned regarding the FileSystemDock should be gone now.
I also fixed a bug where an uncollapsed directory, which is a child of a collapsed directory, got collapsed when you saved the scene (and the Tree is rebuilt). 

`FileSystemDock::_compute_uncollapsed_paths()` is renamed to `FileSystemDock::get_uncollapsed_paths() const` , made public (so it can be used in EditorNode) and uses the new BFS algorithm (which does not include non-directory paths and fixes the bug described above).

In addition, I introduced `get_prev_in_tree(bool p_wrap)` and `get_next_in_tree(bool p_wrap)` (together with some minor internal refactoring).
These were necessary because the TreeItem API lacked a method like `get_next_visible` for iterating over the tree via a DFS, which includes all items, not just the visible. `get_next` just returns a TreeItem's sibling.

(As an alternative approach, we could add two parameters (default: false) to `get_next`, one for changing the mode to DFS and another for wrapping)

KoBeWi commented on doc/classes/TreeItem.xml:
```diff
@@ -246,6 +254,12 @@
+			<return type="TreeItem" />
+			<param index="0" name="wrap" type="bool" default="false" />
+			<description>
+			</description>
```
Needs description. You can copy-paste it from `get_next_in_tree`.
Also I now noticed typo in method's name.

Geometror replied:
Oh, looks like I accidentally inserted that t in the binding name later since I remember writing this description. Fixed.

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -5090,24 +5106,143 @@ void EditorNode::_load_docks_from_config(Ref<ConfigFile> p_layout, const String
+		PackedStringArray uncollapsed_tis = p_layout->get_value(p_section, "dock_filesystem_uncollapsed_paths");
+		for (int i = 0; i < uncollapsed_tis.size(); i++) {
+			TreeItem *uncollapsed_ti = FileSystemDock::get_singleton()->get_tree_control()->get_item_with_metadata(uncollapsed_tis[i], 0);
+			print_line("ti: " + uncollapsed_tis[i]);
+			if (uncollapsed_ti) {
+				print_line("uncollapsed_ti: " + String(uncollapsed_ti->get_metadata(0)));
```
```suggestion
			if (uncollapsed_ti) {
```

KoBeWi commented:
Floating docks are no longer restored properly. They are saved both as floating and in dock position, which makes them non-floating when layout is loaded.

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -196,24 +195,28 @@ Vector<String> FileSystemDock::_compute_uncollapsed_paths() {
-					while (child) {
-						needs_check.push_back(child);
-						child = child->get_next();
+			while (queue.size()) {
```
```suggestion
			while (!queue.is_empty()) {
```

KoBeWi commented on doc/classes/TreeItem.xml:
```diff
@@ -226,11 +226,19 @@
+			<return type="TreeItem" />
+			<param index="0" name="wrap" type="bool" default="false" />
+			<description>
+				Returns the next TreeItem in the tree (in the context of a Depth-first search) or a null object if there is none.
```
```suggestion
				Returns the next TreeItem in the tree (in the context of a depth-first search) or a null object if there is none.
```
Same below.

KoBeWi replied:
This was not fully addressed. While `TreeItem` without braces is ok as is, depth-first search should not be capitalized.

Geometror replied:
Fixed.

KoBeWi commented on doc/classes/TreeItem.xml:
```diff
@@ -226,11 +226,19 @@
+			<param index="0" name="wrap" type="bool" default="false" />
+			<description>
+				Returns the next TreeItem in the tree (in the context of a Depth-first search) or a null object if there is none.
+				If [param wrap] is enabled, the method will wrap around to the first visible element in the tree when called on the last element, otherwise it returns [code]null[/code].
```
"first visible" does not apply here, no?

Geometror replied:
This was a copy-paste mistake.

KoBeWi commented on editor/filesystem_dock.h:
```diff
@@ -125,7 +125,7 @@ class FileSystemDock : public VBoxContainer {
 	Button *button_hist_next = nullptr;
 	Button *button_hist_prev = nullptr;
-	LineEdit *current_path = nullptr;
+	LineEdit *current_path_lineedit = nullptr;
```
```suggestion
	LineEdit *current_path_line_edit = nullptr;
```

KoBeWi commented on editor/filesystem_dock.h:
```diff
@@ -295,7 +294,7 @@ class FileSystemDock : public VBoxContainer {
 	void _search(EditorFileSystemDirectory *p_path, List<FileInfo> *matches, int p_max_items);
 
-	void _set_current_path_text(const String &p_path);
+	void _set_current_path_lineedit_text(const String &p_path);
```
```suggestion
	void _set_current_path_line_edit_text(const String &p_path);
```

KoBeWi commented on editor/filesystem_dock.cpp:
```diff
@@ -196,24 +195,28 @@ Vector<String> FileSystemDock::_compute_uncollapsed_paths() {
 			}
 		}
 	}
+
```
Such blank lines are unnecessary, because the lone `}` already acts as separator.

Geometror commented:
Floating docks should now be restored correctly again, somehow a small code section got automatically removed during the rebase. Addressed all other review comments too.

KoBeWi reviewed (approved):
There's one last documentation comment left, but other than that I think it's fine now.

Geometror committed:
Improve editor state persistence

akien-mga commented:
Thanks!
